docs/architectCustomTests.md

```diff
@@ -0,0 +1,310 @@
+### How to construct custom tests
+
+
+This document breaks down code snippets in [learn_ibc_test.go](../examples/learn_ibc_test.go). Follow along there for a better learning experience.
+
+[learn_ibc_test.go](../examples/learn_ibc_test.go) is a basic test that:
+
+1) Spins up two chains (Gaia and Osmosis) 
+2) Creates an IBC Path between them (client, connection, channel)
+3) Sends an IBC transaction between them.
+
+It validates each step and confirms that the balances of each wallet are correct. 
+
+
+## Chain Factory
+
+```go
+cf := ibctest.NewBuiltinChainFactory(zaptest.NewLogger(t), []*ibctest.ChainSpec{
+    {Name: "gaia", Version: "v7.0.1", ChainConfig: ibc.ChainConfig{
+        GasPrices: "0.0uatom",
+    }},
+    {Name: "osmosis", Version: "v11.0.1"},
+})
+```
+
+The chain factory is where you configure your chain binaries. 
+
+`ibctest` needs a docker image with the chain binary(s) installed to spin up the local testnet. 
+
+`ibctest` has several [pre-configured chains](./preconfiguredChains.txt). These docker images are pulled from [Heighliner](https://github.com/strangelove-ventures/heighliner) (repository of docker images of many IBC enabled chains). Note that Heighliner needs to have the `Version` you are requesting.
+
+When creating your `ChainFactory`,
+If the `Name` matches the name of a pre-configured chain, the pre-configured settings are used. You can override these settings by passing them into the `ibc.ChainConfig` when initializing your ChainFactory. We do this above with the `GasPrices` for gaia above.
+
+You can also pass in **remote images** and/or **local docker images**. 
+
+See an example of each below:
+
+```go
+cf := ibctest.NewBuiltinChainFactory(zaptest.NewLogger(t), []*ibctest.ChainSpec{
+    // PRE CONFIGURED CHAIN EXAMPLE -- gaia chain
+    {Name: "gaia", Version: "v7.0.2"},
+
+    // REMOTE IMAGE EXAMPLE -- ibc-go simd chain
+    {ChainConfig: ibc.ChainConfig{
+        Type: "cosmos",
+        Name: "ibc-go-simd",
+        ChainID: "simd-1",
+        Images: []ibc.DockerImage{
+            {
+                Repository: "ghcr.io/cosmos/ibc-go-simd-e2e",
+                Version: "pr-1973",
+            },
+        },
+        Bin: "simd",
+        Bech32Prefix: "cosmos",
+        Denom: "gos",
+        GasPrices: "0.00gos",
+        GasAdjustment: 1.3,
+        TrustingPeriod: "508h",
+        NoHostMount: false},
+    },
+
+    // LOCAL DOCKER IMAGE EXAMPLE
+    {ChainConfig: ibc.ChainConfig{
+        Type: "cosmos",
+        Name: "stringChain",
+        ChainID: "string-1",
+        Images: []ibc.DockerImage{
+            {
+                Repository: "string", // local docker image name
+                Version: "v1.0.0",	// docker tag 
+            },
+        },
+        Bin: "stringd",
+        Bech32Prefix: "cosmos",
+        Denom: "cheese",
+        GasPrices: "0.01cheese",
+        GasAdjustment: 1.3,
+        TrustingPeriod: "508h",
+        NoHostMount: false},
+    },
+    })
```

Comment:
I appreciate all the code samples in the markdown here. Unfortunately, they risk becoming stale and inaccurate quickly. And there's nothing besides developer discipline reminding us to update this document.

I recommend moving this document's content as comments in `examples/learn_ibc_test.go`. Because `learn_ibc_test.go` is actual code, we'll be forced to keep it up to date as ibctest gets extended and refactored. Then this document becomes sparse, perhaps laying out some high level detail and informing the user to check out the example test.

Reply:
I partially disagree here. 
Since this is a doc on how to write custom tests, I think the code snippets help break up the document making it easier to read and give context to what a "Chain Factory", "Interchain", ect. look like.
As for the code getting stale... I really did try and keep this high level. But maybe I am taking for granted the rate of refactors!?

Maybe there are too many code snippets? There's a few extras in here just to show the learner that there are multiple ways to do the same thing, some of which are 1) hard to discover and 2) important for the writer to know depending on what they are testing.

For me personally, this Is how I learn best — having a script that is complete, and then having a description of each section. I think it would be a much tougher read if this was commented out in a go file. Given the audience, perhaps this is too wordy and unnecessary?

I removed some code and condensed a little in the latest push. 
If consensus is to make this less code and more high level explanations, I’m happy to make more adjustments. Regardless, appreciate any/all feedback.

Reply:
Go also has a feature for [testable examples](https://go.dev/blog/examples) we may be able to leverage. I'm not sure if that's a good fit there, but it's another option - perhaps for a future PR.